docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
```

Comment:
The reverse proxy approach can work (we use it in OpenShift today to
authorize / authenticate kube under a single namespace, see
https://github.com/openshift/origin/blob/b7de777c90fd892cd746c905c17d92710eae3de8/pkg/cmd/server/kubernetes/proxy.go#L17),
i just don't want to hobble this proposal to assume its existence.

On Tue, Nov 17, 2015 at 7:13 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> An example is kerberos security, where you want the client to access
> individual servers and get delegation.  Reverse proxy sets up a point of
> control inversion where the core master is a proxy mitm for all clients.
> That works fine in some environments, but runs afoul of security practices
> on others.
>
> On Tue, Nov 17, 2015 at 7:09 PM, Clayton Coleman <ccoleman@redhat.com>
> wrote:
>
>> We have to do autonomous client discovery.  In the other issue I
>> mentioned that we don't want to proxy all possible extensions through the
>> master - I think that still stands.  Clients in some scenarios have to be
>> able to talk to multiple servers.  I would love the option to support
>> reverse proxy, but I don't think this proposal requires it.
>>
>> On Tue, Nov 17, 2015 at 7:02 PM, Prashanth B <notifications@github.com>
>> wrote:
>>
>>> In docs/proposals/third-party-api-servers.md
>>> <https://github.com/kubernetes/kubernetes/pull/17191#discussion_r45141648>
>>> :
>>>
>>> > +[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
>>> > +
>>> > +Documentation for other releases can be found at
>>> > +[releases.k8s.io](http://releases.k8s.io).
>>> > +</strong>
>>> > +--
>>> > +
>>> > +<!-- END STRIP_FOR_RELEASE -->
>>> > +
>>> > +<!-- END MUNGE: UNVERSIONED_WARNING -->
>>> > +
>>> > +# Third Party API Servers
>>> > +
>>> > +## Goal
>>> > +
>>> > +With third party API servers, we want to enable cluster admins or users to
>>>
>>> Hmm, the way I read your comment you're asking that we run a haproxy (or
>>> something) in front of each third party server. Isn't this a pain?
>>>
>>> This one paragraph is reminiscent of all the router/serviceloadbalancer
>>> stuff we did. Some services come up and register themselves with a single
>>> go program+haproxy capable of programming itself dynamically. I think this
>>> could be useful even today (without third party apiservers), because we'd
>>> just spin up 2 apiservers fronted by haproxy.
>>>
>>> We can give that router go program a "dumb" backend that doesn't need
>>> haproxy at all of course.
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/kubernetes/kubernetes/pull/17191/files#r45141648>.
>>>
>>
>>
>